### PR TITLE
Add close state to tsdb.DB

### PR DESCRIPTION
**Pull request by damnever**

### CODE REVIEW

**Comment by codesome** [comment]
Does this solve any known issue?

**Comment by damnever** [comment]
> Does this solve any known issue?

No, but a more consistent state makes this "library" easier to use, for example, return an error or nil is better than panic when calling `DB.Close` twice.

**Comment by damnever** [comment]
I have pushed another [commit](https://github.com/prometheus/prometheus/pull/11586/commits/ddc386b066be94af869478851129f9fcf78d0b39) to fix the logic when open tsdb.DB, and it may cause a panic now without the close state.

**Comment by bboreham** [comment]
My opinion is not to add more code except when fixing a bug, adding a feature that users have asked for, or clearly improving performance.

**Comment by damnever** [comment]
@bboreham If you read my comment carefully, this is a BUG fix:
> 

> I have pushed another [commit](https://github.com/prometheus/prometheus/pull/11586/commits/ddc386b066be94af869478851129f9fcf78d0b39) to fix the logic when open tsdb.DB, and it may cause a panic now without the close state.

File: tsdb/db.go [hunk 1]
```diff
@@ -657,7 +658,7 @@ func validateOpts(opts *Options, rngs []int64) (*Options, []int64) {
 // open returns a new DB in the given directory.
 // It initializes the lockfile, WAL, compactor, and Head (by replaying the WAL), and runs the database.
 // It is not safe to open more than one DB in the same directory.
-func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs []int64, stats *DBStats) (_ *DB, returnedErr error) {
```
Comment (damnever) [inline tsdb/db.go:660]: @bboreham 
`returnedErr` is always nil before this fix.

Comment (damnever) [reply tsdb/db.go:660]: I have tested it, it is not..
```go
package main

import "fmt"

func foo() (err error) {
	defer func() {
		fmt.Println(err)
	}()
	return fmt.Errorf("oops")
}

func main() {
	foo()
}
```
magic!

**Pull request closed without merging**